New User entity added

jszymanski97 opened the pull request:
Model, DTOs, mapper, service, repository and controller created for new entity: User

New endpoint allows for registering new user in to database.

jszymanski97 committed:
Model, DTOs, mapper, service, repository and controller created for new entity: User
New endpoint allows for registering new user in to database.

slade13 commented on src/main/java/mate/project/dto/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,33 @@
+            message = "Password must contain at least one uppercase letter,"
+                    + " one digit, and one special character.")
+    private String password;
+    private String repeatPassword;
```
The 'repeatPassword' field is not validated. It's recommended to add validation annotations to ensure data integrity.

slade13 commented on src/main/java/mate/project/dto/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,33 @@
+    @NotNull
+    @Size(min = 1, max = 255)
+    private String lastName;
+    private String shippingAddress;
```
The 'shippingAddress' field is not validated. Consider adding validation annotations or provide a comment if it's intentionally unvalidated.

slade13 commented on src/main/java/mate/project/controller/AuthenticationController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @PostMapping("/registration")
+    public UserResponseDto register(@Valid @RequestBody UserRegistrationRequestDto request)
+            throws RegistrationException {
```
The method 'register' in the controller should not throw exceptions. Instead, handle the exception within the method and return a suitable ResponseEntity as response.

slade13 commented on src/main/java/mate/project/mapper/UserMapper.java:
```diff
@@ -0,0 +1,16 @@
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
```
The @Mapper annotation should include componentModel = 'spring' for the UserMapper to be recognized as a Spring bean.

slade13 commented on src/main/java/mate/project/mapper/UserMapper.java:
```diff
@@ -0,0 +1,16 @@
+    UserResponseDto toUserResponseDto(User user);
+
+    @Mapping(target = "id", ignore = true)
+    User toModel(UserRegistrationRequestDto userRegistrationRequestDto);
```
The @Mapping annotation is used to ignore the 'id' field when mapping from UserRegistrationRequestDto to User. Given that UserRegistrationRequestDto likely doesn't have an 'id' field, this annotation may not be necessary.

jszymanski97 replied:
Without having this @Mapping annotation here maven is keep throwing me issues that field id is not mapped

slade13 commented on src/main/java/mate/project/validation/FieldMatchValidator.java:
```diff
@@ -0,0 +1,24 @@
+    public boolean isValid(Object object, ConstraintValidatorContext context) {
+        Object firstValue = new BeanWrapperImpl(object).getPropertyValue(first);
+        Object secondValue = new BeanWrapperImpl(object).getPropertyValue(second);
+        return Objects.equals(firstValue, secondValue);
```
The Objects.equals method will handle null values, but it may be better to explicitly handle the case where both fields are null. Depending on the requirements, two null values may not be considered a match.

slade13 commented on src/main/java/mate/project/validation/FieldMatchValidator.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public boolean isValid(Object object, ConstraintValidatorContext context) {
+        Object firstValue = new BeanWrapperImpl(object).getPropertyValue(first);
```
The BeanWrapperImpl instances are being created twice, which is a performance overhead. Consider creating a single instance of BeanWrapperImpl and reusing it.

slade13 commented on src/main/java/mate/project/validation/FieldMatchValidator.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.beans.BeanWrapperImpl;
+
+public class FieldMatchValidator implements ConstraintValidator<FieldMatch, Object> {
+    private String first;
```
The 'first' and 'second' instance variables could have more descriptive names to improve code readability.

slade13 commented on src/main/java/mate/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()) != null) {
```
The check for an existing user can lead to a race condition where two users with the same email can be created at the same time. Consider improving the uniqueness check at the database level, not at the service level.

jszymanski97 replied:
I thought the unique = true annotation for email in User model class also makes sure that the email is unique at the database level throwing the DataIntegrityViolationException which I now have also included to be handled in UserService hope this is a good solution

slade13 commented on src/main/java/mate/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
```
Consider not using checked exceptions for business logic, use unchecked/runtime exceptions instead.

slade13 commented on src/main/java/mate/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+                    + " already exists");
+        }
+        User model = userMapper.toModel(requestDto);
+        userRepository.save(model);
```
It's not clear if the 'save' operation is successful or not as no result of the operation is being checked or used. Consider checking the result of the 'save' operation.

slade13 commented on src/main/java/mate/project/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -48,6 +48,18 @@ public ResponseEntity<Object> handleEntityNotFoundException(
+            RegistrationException ex,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", ZonedDateTime.now());
```
The timestamp is set with the current timezone, which may lead to confusion if users are in different timezones. It would be better to use UTC to avoid this confusion.

slade13 commented on src/main/java/mate/project/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -48,6 +48,18 @@ public ResponseEntity<Object> handleEntityNotFoundException(
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", ZonedDateTime.now());
+        body.put("message", ex.getMessage());
+        body.put("status", HttpStatus.BAD_REQUEST.value());
```
The HTTP status code is hardcoded. It would be better to get the status code from the exception, if possible.

jszymanski97 replied:
I am not sure why I am getting comments on code that has been already approved by other mentors? Especially that it doesn't relate to the new functions that have been added in the pull request

slade13 commented on src/main/java/mate/project/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -48,6 +48,18 @@ public ResponseEntity<Object> handleEntityNotFoundException(
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", ZonedDateTime.now());
+        body.put("message", ex.getMessage());
```
The error message is directly from the exception. It would be better to create a user-friendly message.

slade13 commented on src/main/java/mate/project/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -48,6 +48,18 @@ public ResponseEntity<Object> handleEntityNotFoundException(
+        body.put("message", ex.getMessage());
+        body.put("status", HttpStatus.BAD_REQUEST.value());
+
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
The error response is not fully descriptive. It is recommended to include additional information such as path, method, and parameters in the error response.

slade13 commented on src/main/java/mate/project/validation/FieldMatch.java:
```diff
@@ -0,0 +1,20 @@
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface FieldMatch {
+    String message() default "Fields do not match";
```
The default error message 'Fields do not match' is quite generic. It might be more helpful to users if the message specified which fields did not match.

slade13 commented on src/main/java/mate/project/validation/FieldMatch.java:
```diff
@@ -0,0 +1,20 @@
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+
+    String first();
```
There is no documentation or comments indicating what 'first()' and 'second()' methods are supposed to represent. It might be helpful to provide documentation for these methods to indicate their purpose.

slade13 commented on src/main/java/mate/project/model/User.java:
```diff
@@ -0,0 +1,32 @@
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    private String shippingAddress;
```
The 'shippingAddress' field should be annotated with @Column. Even though it's optional, it's a good practice to include it for clarity and consistency.

slade13 commented on src/main/java/mate/project/model/User.java:
```diff
@@ -0,0 +1,32 @@
+    @Column(nullable = false, unique = true)
+    private String email;
+    @Column(nullable = false)
+    private String password;
```
The 'password' field should be encrypted and not stored as plain text for security reasons. It's best to use encryption methods or hashing algorithms to store passwords.

jszymanski97 replied:
I am not too sure how to implment this yet, I believe in further lessons there is a video and theory explaining this, I will make sure to change it then

slade13 commented on src/main/java/mate/project/dto/UserResponseDto.java:
This DTO doesn't have any data validation

jszymanski97 replied:
What sort of validations would a DTO have that is only used to return data? I have researched in to this and forums say that generally you are not adding validations to DTOs used only for returning

slade13 reviewed (changes requested):
Please review my comments and update your pull request accordingly.

jszymanski97 committed:
UserRegistrationRequestDto - Extra validations added,
FieldMatchValidator - enhanced handling of null values,
UserServiceImpl - now handles the RegistrationException,
CustomGlobalExceptionHandler enhanced return of messages and status code

slade13 commented on src/main/java/mate/project/dto/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,38 @@
+            message = "Password must contain at least one uppercase letter,"
+                    + " one digit, and one special character.")
+    private String password;
+    @Size(min = 8, max = 255)
```
Validation annotation `@NotNull` is missing for the field 'repeatPassword'. It should be there to avoid null values and to maintain consistency with other fields.

slade13 commented on src/main/java/mate/project/exception/RegistrationException.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+public class RegistrationException extends RuntimeException {
+    private final HttpStatus status;
```
The variable 'status' should not be part of the exception. Exceptions should not know anything about HTTP statuses.

slade13 reviewed (approved):
Good job on updating those comments, please look on my new notes which can help you to improve your code.

jszymanski97 committed:
Removed Http status from custom exception
